packages/terra-application-utility/src/utility/_UtilityMenu.jsx

```diff
@@ -0,0 +1,319 @@
+import Button from 'terra-button';
+import ContentContainer from 'terra-content-container';
+import IconClose from 'terra-icon/lib/icon/IconClose';
+import IconLeft from 'terra-icon/lib/icon/IconLeft';
+import MenuDivider from './_UtilityMenuDivider';
+import Utils from '../_Utils';
+import MenuItem from './_UtilityMenuItem';
+import styles from './_UtilityMenu.scss';
+
+const cx = classNames.bind(styles);
+
+const propTypes = {
+  /**
+   * The initial selected key. Used as the top level menu page.
+   */
+  selectedKey: PropTypes.string.isRequired,
+  /**
+   * Indicates if the height is bound to it's parent container.
+   */
+  isHeightBounded: PropTypes.bool,
+  /**
+   * The array of the menu items configs to be rendered.
+   */
+  menuItems: PropTypes.arrayOf((Utils.itemShape)).isRequired,
+  /**
+   * The function to trigger when a menu item is selected.
+   */
+  onChange: PropTypes.func.isRequired,
+  /**
+   * The function that closes the menu.
+   * This will be provided by the terra-application-header or terra-application-menu
+   */
+  onRequestClose: PropTypes.func,
+  /**
+   * Sets the Utility variant. One of Utils.VARIANTS.HEADER, Utils.VARIANTS.MENU.
+   */
+  variant: PropTypes.oneOf([Utils.VARIANTS.HEADER, Utils.VARIANTS.MENU]),
+};
+
+const processMenuItems = (items) => {
+  const map = new Map();
+  items.forEach((item) => {
+    map.set(
+      item.key,
+      { id: item.id,
+        itemKey: item.key,
+        title: item.title,
+        content: item.content,
+        contentLocation: item.contentLocation,
+        isSelected: item.isSelected,
+        isSelectable: item.isSelectable,
+        childKeys: item.childKeys,
+      },
+    );
+  });
+  return map;
+};
+
+const hasChevron = item => item.childKeys && item.childKeys.length > 0;
+
+const contextTypes = {
+  /* eslint-disable consistent-return */
+  intl: (context) => {
+    if (context.intl === undefined) {
+      return new Error('Please add locale prop to Base component to load translations');
+    }
+  },
+};
+
+class UtilityMenu extends React.Component {
+  constructor(props) {
+    super(props);
+    this.getItem = this.getItem.bind(this);
+    this.buildItem = this.buildItem.bind(this);
+    this.buildListContent = this.buildListContent.bind(this);
+    this.buildFooterContent = this.buildFooterContent.bind(this);
+    this.childrenHasCheckmark = this.childrenHasCheckmark.bind(this);
+    this.childrenHasChevron = this.childrenHasChevron.bind(this);
+    this.handleOnChange = this.handleOnChange.bind(this);
+    this.handleRequestBack = this.handleRequestBack.bind(this);
+    this.pop = this.pop.bind(this);
+    this.push = this.push.bind(this);
+    this.state = {
+      map: processMenuItems(props.menuItems),
+      currentKey: props.selectedKey,
+      previousKeyStack: [],
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.menuItems === this.props.menuItems) {
+      return;
+    }
+
+    this.setState({
```

Comment:
If the selectedKey prop changes from one update to the next, we should reset the currentKey state value to reflect that update. 

This is a bit of an edge case, and I doubt it'd happen often, but if consumers are changing that prop we should account for it.

Reply:
Do we want to rename it to rootKey or something then?

Reply:
Yeah, we went ahead with `initialMenuKey`